### PR TITLE
Enforce AttributeTargets.Class and AttributeTargets.Struct

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

Fixes https://github.com/dotnet/fsharp/issues/13526

### Old rule
- `Classes and Struct` were using `AttributeTargets.Class ||| AttributeTargets.Interface ||| AttributeTargets.Delegate ||| AttributeTargets.Struct ||| AttributeTargets.Enum`

### New rule
- Classes use `AttributeTargets.Class`
- Structs use `AttributeTargets.Struct`

### Before
```fsharp
open System

[<AttributeUsage(AttributeTargets.Class)>]
type CustomClassAttribute() =
    inherit Attribute()

[<AttributeUsage(AttributeTargets.Struct)>]
type CustomStructAttribute() =
    inherit Attribute()

[<CustomClass; CustomStruct>] // No Error
type Class() = class end

[<CustomClass; CustomStruct>] // No Error
type Struct(x: int) = struct end

[<CustomClass>] // No Error
[<Struct>]
type Struct2(x: int) = ...

[<CustomStruct>] // No Error
[<Class>]
type Class2() = ...

```

### After
```fsharp
open System

[<AttributeUsage(AttributeTargets.Class)>]
type CustomClassAttribute() =
    inherit Attribute()

[<AttributeUsage(AttributeTargets.Struct)>]
type CustomStructAttribute() =
    inherit Attribute()

[<CustomClass; CustomStruct>] // Error as expected as you can not use AttributeTargets.Struct in a class
type Class() = class end

[<CustomClass; CustomStruct>] // Error as expected as you can not use AttributeTargets.Class in a struct
type Struct(x: int) = struct end

[<CustomClass>] // Error as expected as you can not use AttributeTargets.Class in a struct
[<Struct>]
type Struct2(x: int) = ...

[<CustomStruct>] // Error as expected as you can not use AttributeTargets.Struct in a class
[<Class>]
type Class2() = ...
```

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by edgarfgp** [comment]
This is ready

File: src/Compiler/Service/SemanticClassification.fs [hunk 1]
```diff
@@ -58,7 +58,6 @@ type SemanticClassificationType =
     | TypeDef = 35
     | Plaintext = 36
 
-[<RequireQualifiedAccess>]
```
Comment (edgarfgp) [inline src/Compiler/Service/SemanticClassification.fs:61]: `RequireQualifiedAccess` is defined as AttributeTargets.Class so it should be used here

File: src/Compiler/Service/SemanticClassification.fsi [hunk 2]
```diff
@@ -49,7 +49,6 @@ type SemanticClassificationType =
     | TypeDef = 35
     | Plaintext = 36
 
-[<RequireQualifiedAccess>]
```
Comment (edgarfgp) [inline src/Compiler/Service/SemanticClassification.fsi:52]: `RequireQualifiedAccess` is defined as AttributeTargets.Class so it should be used here

**Comment by vzarytovskii** [comment]
This will likely break the following scenario:

```fsharp
[<Struct; RequireQualifiedAccess>]
type x = { y: int }
let xx = { x.y = 123 } // Works and will emit "The record label 'y' is not defined" diagnostic without "x."
```

Will it now throw `FS842`? In case if it will, we probably can't have it as an error by default, too many things may break and either a bunch of fixes will be needed to all sorts of attributes in libraries or rewrites of the underlying types.

**Comment by edgarfgp** [comment]
> This will likely break the following scenario:
> 
> ```fsharp
> [<Struct; RequireQualifiedAccess>]
> type x = { y: int }
> let xx = { x.y = 123 } // Works and will emit "The record label 'y' is not defined" diagnostic without "x."
> ```
> 
> Will it now throw `FS842`? In case if it will, we probably can't have it as an error by default, too many things may break and either a bunch of fixes will be needed to all sorts of attributes in libraries or rewrites of the underlying types.

This is unaffected as this PR checks for `SynTypeDefnSimpleRepr.General` as opposed `SynTypeDefnSimpleRepr.Record`

**Comment by vzarytovskii** [comment]
> > This will likely break the following scenario:
> > ```fsharp
> > [<Struct; RequireQualifiedAccess>]
> > type x = { y: int }
> > let xx = { x.y = 123 } // Works and will emit "The record label 'y' is not defined" diagnostic without "x."
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Will it now throw `FS842`? In case if it will, we probably can't have it as an error by default, too many things may break and either a bunch of fixes will be needed to all sorts of attributes in libraries or rewrites of the underlying types.
> 
> This is unaffected as this PR checks for `SynTypeDefnSimpleRepr.General` as opposed `SynTypeDefnSimpleRepr.Record`

Ah, good.

**Comment by vzarytovskii** [comment]
What about something like
```fsharp

<RequireQualifiedAccess; Struct; NoComparison>]
type ValueOrCancelled<'TResult> =
    | Value of result: 'TResult
    | Cancelled of ``exception``: OperationCanceledException
```

There might be a bunch of similar cases, with rqa in particular. At the very least we should extend the attribute to structs (possibly other targets).

**Comment by edgarfgp** [comment]
> What about something like
> 
> ```fsharp
> <RequireQualifiedAccess; Struct; NoComparison>]
> type ValueOrCancelled<'TResult> =
>     | Value of result: 'TResult
>     | Cancelled of ``exception``: OperationCanceledException
> ```
> 
> There might be a bunch of similar cases, with rqa in particular. At the very least we should extend the attribute to structs (possibly other targets).

This is also unaffected as this is PR does check `SynTypeDefnSimpleRepr.Union` and we are only checking for `SynTypeDefnSimpleRepr.General`

- `SynTypeDefnSimpleRepr.General` will cover cases like:

```fsharp

type MyClass() = class end

[<Class>]
type MyClass = ...

type MyStruct(x: int) = struct end

[<Struct>]
type MyStruct(x: int) = ...
```

So records, union and it generic counterparts still works as before.


Currently `RequireQualifiedAccessAttribute` is defined as follows. But I guess we can add `AttributeTargets.Struct` too
```fsharp
    [<AttributeUsage(AttributeTargets.Class, AllowMultiple=false)>]
    [<Sealed>]
    type RequireQualifiedAccessAttribute() =
        inherit Attribute()
```

**Comment by edgarfgp** [comment]
@auduchinok Could you have a look this one please ?

**Comment by KevinRansom** [comment]
This feels more like a bug fix than a language feature, does it really need to be tied to a language version flag?

**Comment by edgarfgp** [comment]
> This feels more like a bug fix than a language feature, does it really need to be tied to a language version flag?

Agreed it is a bug, But based on previous PR's on AttributeTargets this code will stop compiling in the next version so we do not want to break existing code, that being said happy to remove the language version flag

File: src/Compiler/Service/SemanticClassification.fs (hunk 1 above)
Comment (T-Gro) [reply src/Compiler/Service/SemanticClassification.fs:61]: @edgarfgp we do need the RQA here, it is one of the way to keep sanity when working in the compiler codebase.
I think with this bug you found the more appropriate action is to extend RQA to also allow structs.

Also, struct-based DUs are the same story - we definitely want to keep allowing RQA.

File: tests/service/data/TestTP/ProvidedTypes.fs [hunk 3]
```diff
@@ -3220,8 +3220,6 @@ module internal AssemblyReader =
           systemRuntimeScopeRef: ILScopeRef }
         override __.ToString() = "<ILGlobals>"
 
-    [<AutoOpen>]
```
Comment (T-Gro) [inline tests/service/data/TestTP/ProvidedTypes.fs:3223]: AutoOpen is the same story - rather then removing it, it makes more sense to extend it also on structs.

**Review by T-Gro — CHANGES_REQUESTED** [review]
Nice finding @edgarfgp .
See the other comment - I would rather go by extending RQA and AutoOpen to also cover structs, instead of banning it.

**Comment by T-Gro** [comment]
The steps I would recommend:

- find all usages of "AttributeTargets.Class" in FSharp.Core.
foreach usage do:
   if it also has AttributeTargets.Struct or AttributeTargets.All, continue;
   if it is clearly relevant only in class context (abstract, sealed, ..), continue;
   if it is relevant only in F#-special structs not touched by this PR (DU augmentations, records equality, etc.) - keep in mind for next phases, continue;

   add AttributeTargets.Struct to the list of allowed targets.

File: src/Compiler/Service/SemanticClassification.fsi (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/Service/SemanticClassification.fsi]: Reverted as we extended `RequireQualifiedAccess` to use `AttributeTargets.Struct`

**Comment by vzarytovskii** [comment]
> This feels more like a bug fix than a language feature, does it really need to be tied to a language version flag?

It is a breaking change, hence should go under language flag.

**Comment by edgarfgp** [comment]
@vzarytovskii the CI seems to stuck at `Waiting for status to be reported`

**Comment by vzarytovskii** [comment]
> @vzarytovskii the CI seems to stuck at `Waiting for status to be reported` 

We do some maintenance for separating PR and signed builds. It won't work, I posted message on Slack some time ago.

**Comment by edgarfgp** [comment]
> The steps I would recommend:
> 
> * find all usages of "AttributeTargets.Class" in FSharp.Core.
>   foreach usage do:
>   if it also has AttributeTargets.Struct or AttributeTargets.All, continue;
>   if it is clearly relevant only in class context (abstract, sealed, ..), continue;
>   if it is relevant only in F#-special structs not touched by this PR (DU augmentations, records equality, etc.) - keep in mind for next phases, continue;
>   add AttributeTargets.Struct to the list of allowed targets.

@T-Gro I have added `AttributeTargets.Struct` to `AutoOpen` and `RequireQualifiedAccess`  

Also had a look at the places where `AttributeTargets.Class` is used in FSharp.Core and did not find any clear candidate for adding also `AttributeTargets.Struct`.

**Comment by vzarytovskii** [comment]
> The steps I would recommend:
> 
> * find all usages of "AttributeTargets.Class" in FSharp.Core.
>   foreach usage do:
>   if it also has AttributeTargets.Struct or AttributeTargets.All, continue;
>   if it is clearly relevant only in class context (abstract, sealed, ..), continue;
>   if it is relevant only in F#-special structs not touched by this PR (DU augmentations, records equality, etc.) - keep in mind for next phases, continue;
>   add AttributeTargets.Struct to the list of allowed targets.

Shall Enum target also be added?

**Comment by edgarfgp** [comment]
> > The steps I would recommend:
> > 
> > * find all usages of "AttributeTargets.Class" in FSharp.Core.
> >   foreach usage do:
> >   if it also has AttributeTargets.Struct or AttributeTargets.All, continue;
> >   if it is clearly relevant only in class context (abstract, sealed, ..), continue;
> >   if it is relevant only in F#-special structs not touched by this PR (DU augmentations, records equality, etc.) - keep in mind for next phases, continue;
> >   add AttributeTargets.Struct to the list of allowed targets.
> 
> Shall Enum target also be added?

I think it makes sense in the context of `AutoOpen` and `RequireQualifiedAccess` as then can be used in `classes, structs and enums`. 

But we still need to enforce `AttributeTargets.Enum` only `enums` [sharplab](https://sharplab.io/#v2:DYLgZgzgPgsAUAewA4FMB2ACAygTwgFxQFt558dUMANARgwF54NmoMBBBjABiA==) which I would cover in a separate PR

**Comment by edgarfgp** [comment]
Can I get another review here please ?

File: docs/release-notes/.FSharp.Core/8.0.300.md [hunk 4]
```diff
@@ -6,3 +6,4 @@
 ### Fixed
 
 * Preserve original stack traces in resumable state machines generated code if available. ([PR #16568](https://github.com/dotnet/fsharp/pull/16568))
+* Enforce AttributeTargets on structs and classes ([PR #16790](https://github.com/dotnet/fsharp/pull/16790))
```
Comment (vzarytovskii) [inline docs/release-notes/.FSharp.Core/8.0.300.md:9]: I think this should say that we extended `RequireQualifiedAccessAttribute` and `AutoOpenAttribute` to work on structs as well.

**Review by vzarytovskii — APPROVED** [review]

File: docs/release-notes/.FSharp.Core/8.0.300.md (hunk 4 above)
Comment (edgarfgp) [reply docs/release-notes/.FSharp.Core/8.0.300.md]: Done

**Comment by edgarfgp** [comment]
Closing/Reopening to trigger the CI again :(

**Review by vzarytovskii — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]
Well done and well tested. Thanks!
@T-Gro?

**Comment by T-Gro** [comment]
On top of RQA and AutoOpen, attributes where I think this change can potentially introduce breaks in user code:

- AutoSerializableAttribute
- StructuralComparisonAttribute
- StructuralEqualityAttribute
- ReflectedDefinitionAttribute  (e.g. on a struct without fields and with static members)

(I am omitting all attributes working on DUs/records, as this is not the scope of this PR - but might be in the scope of future ones)

**Comment by edgarfgp** [comment]
> * ReflectedDefinitionAttribute

@T-Gro Yeah I think the mentioned attributes should updated to target `AttributeTargets.Struct` or`AttributeTargets.Class` when needed. It would be good to do that in a separate PR ?

**Comment by vzarytovskii** [comment]
TypeProviderAttribute also needs to be looked at.

**Comment by edgarfgp** [comment]
> `TypeProviderAttribute` also needs to be looked at.

@vzarytovskii Can you create an issue with all the Attributes that needs updating. Happy to do the actual work

**Comment by vzarytovskii** [comment]
> > `TypeProviderAttribute` also needs to be looked at.
> 
> @vzarytovskii Can you create an issue with all the Attributes that needs updating. Happy to do the actual work

I will update them as part of this PR

**Comment by vzarytovskii** [comment]
> > > `TypeProviderAttribute` also needs to be looked at.
> > 
> > 
> > @vzarytovskii Can you create an issue with all the Attributes that needs updating. Happy to do the actual work
> 
> I will update them as part of this PR, so potentially breaking change won't accidentally leak into merge/release.

Done

**Comment by vzarytovskii** [comment]
> Nice finding @edgarfgp . See the other comment - I would rather go by extending RQA and AutoOpen to also cover structs, instead of banning it.

This is done

**Comment by edgarfgp** [comment]
Thanks @vzarytovskii You rock 🎸

**Comment by vzarytovskii** [comment]
@T-Gro can you re-review pls?

**Review by T-Gro — APPROVED** [review]
Looks good now, thanks for incorporating the changes needed.

**Comment by edgarfgp** [comment]
I guess this PR is taking all merge conflicts today haha

**Comment by edgarfgp** [comment]
Can we have this now in ?

**Pull request merged**